### PR TITLE
Checking Torch inside unit test

### PR DESCRIPTION
**Pull request by plrectco**
## What changes were proposed in this pull request?
1.  Implement a new class TorchSpec extending FlatSpec. It implement the method torchCheck to check for torch. Got this idea from @srowen 
2. Move the torch check inside the unit test

## How was this patch tested?
Local unit tests and style checks.
It shows the test is cancelled when `mvn clean package` on a non-torch machine

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/805

### CODE REVIEW

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/torch/AbsCriterionSpec.scala [hunk 1]
```diff
@@ -20,14 +20,9 @@ import com.intel.analytics.bigdl.tensor.Tensor
...
-  before {
-    if (!TH.hasTorch()) {
-      cancel("Torch is not installed")
-    }
-  }
-
-  "A Abs Criterion " should "generate correct output and grad" in {
+class AbsCriterionSpec extends TorchSpec with BeforeAndAfter with Matchers{
+    "A Abs Criterion " should "generate correct output and grad" in {
+    torchCheck()
```
Comment (srowen) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/torch/AbsCriterionSpec.scala]: Why not just check this in `TorchSpec` with a `beforeAll`? then you don't need this call at all.
`TorchSpec` can extend `BeforeAndAfter`, `BeforeAndAfterAll` and `Matchers` and then all these tests don't have to.

This is the change I was finishing locally and it works fine that way. It still generates an `*** ABORTED ***` message of course.

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/torch/TorchSpec.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.torch
+
+import org.scalatest.FlatSpec
+
+class TorchSpec extends FlatSpec {
+  def torchCheck(): Unit = {
+    if (!TH.hasTorch()) {
```
Comment (srowen) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/torch/TorchSpec.scala]: `assume(TH.hasTorch())` is simpler and does the same thing.

Comment (qiuxin2012) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/torch/TorchSpec.scala]: Output of ```
if (!TH.hasTorch()) {
      cancel("Torch is not installed")
}
```
is something like
```
load torch's Inception+bn float version
- should generate correct output !!! CANCELED !!!
  Torch is not installed (InceptionSpec.scala:607)
```

output of `assume(TH.hasTorch(), "Torch is not installed")` is something like
```
AlexNet float
- should generate correct output !!! CANCELED !!!
  com.intel.analytics.bigdl.torch.TH.hasTorch() was false Torch is not installed! (AlexNetSpec.scala:36)
```
The former is better I think.

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/torch/AbsCriterionSpec.scala (hunk 1 above)
Comment (plrectco) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/torch/AbsCriterionSpec.scala]: I have added `BeforeAndAfter` and `Matchers` to the `TorchSpec`. Is it necessary to add `BeforeAndAfterAll` given we already have `BeforeAndAfter`?

**Comment by plrectco** [comment]
It has passed the Jenkins Test

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/models/AlexNetSpec.scala [hunk 3]
```diff
@@ -22,7 +22,7 @@ import com.intel.analytics.bigdl.nn._
 import com.intel.analytics.bigdl.numeric.NumericFloat
 import com.intel.analytics.bigdl.optim.SGD
 import com.intel.analytics.bigdl.tensor._
-import com.intel.analytics.bigdl.torch.TH
+import com.intel.analytics.bigdl.torch.{TH, TorchSpec}
 import com.intel.analytics.bigdl.utils.RandomGenerator._
 import com.intel.analytics.bigdl.utils.T
 import org.scalatest.{BeforeAndAfter, FlatSpec, Matchers}
```
Comment (srowen) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/models/AlexNetSpec.scala]: You can/should remove these unused imports now too. That's more of the value of the cleanup.

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/torch/TorchSpec.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.torch
+import org.scalatest.{BeforeAndAfter, FlatSpec, Matchers}
+
+class TorchSpec extends FlatSpec with BeforeAndAfter with Matchers {
+  def torchCheck(): Unit = {
```
Comment (srowen) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/torch/TorchSpec.scala]: OK I understand about cancel vs assume, seems fine. But why shouldn't this just be `beforeAll`? then you don't need to call this method everywhere. That's what `BeforeAndAfterAll` is there for.

Comment (plrectco) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/torch/TorchSpec.scala]: I switch to use `beforeAll` but it causes `mvn clean package` to fail. I think `beforeAll` is like the `before`s we currently have for each test .

```
*** 120 SUITES ABORTED ***
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] BigDL .............................................. SUCCESS [  1.064 s]
[INFO] spark_bigdl ........................................ SUCCESS [  0.010 s]
[INFO] spark-version ...................................... SUCCESS [  0.010 s]
[INFO] 1.5-plus ........................................... SUCCESS [  9.350 s]
[INFO] bigdl .............................................. FAILURE [17:16 min]
[INFO] dist-spark-1.5.1-scala-2.10.5-linux64 .............. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 17:27 min
[INFO] Finished at: 2017-05-01T12:23:58+08:00
[INFO] Final Memory: 131M/6454M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.scalatest:scalatest-maven-plugin:1.0:test (test) on project bigdl: There are test failures -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :bigdl

```

Comment (srowen) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/torch/TorchSpec.scala]: Oh I get it now, yes you're right! huh, so cancel/assume only produces the cancelled status (vs aborted) when called directly within the method. Ugh, that's unfortunate. Looks good.

**Comment by srowen** [comment]
@qiuxin2012 @plrectco this looks good. By the way, if this makes the build pass, we can easily enable Travis CI integration to run tests for pull requests automatically. That might save a lot of manual testing for you

**Pull request merged**